evm/db/journal.py

```diff
@@ -174,53 +200,56 @@ def _validate_checkpoint(self, checkpoint):
         """
         Checks to be sure the checkpoint is known by the journal
         """
-        if checkpoint not in self.journal:
+        if not self.journal.has_checkpoint(checkpoint):
             raise ValidationError("Checkpoint not found in journal: {0}".format(
                 str(checkpoint)
             ))
 
-    def snapshot(self):
+    def record(self):
         """
-        Takes a snapshot of the database by creating a checkpoint.
+        Starts a new recording and returns an id for the associated changeset
         """
         return self.journal.create_checkpoint()
 
-    def revert(self, checkpoint):
+    def forget(self, changeset):
         """
-        Reverts the database back to the checkpoint.
+        Throws away all journaled data starting at the given changeset
         """
-        self._validate_checkpoint(checkpoint)
-
-        for key, value in self.journal.pop_checkpoint(checkpoint).items():
-            if value is None:
-                self.wrapped_db.delete(key)
-            else:
-                self.wrapped_db.set(key, value)
+        self._validate_checkpoint(changeset)
+        self.journal.pop_checkpoint(changeset)
 
     def commit(self, checkpoint):
         """
-        Commits a given checkpoint.
+        Commits a given checkpoint. This involves committing the journal up to
+        the given checkpoint which returns all of the journal changes.  *if*
+        this is the base checkpoint then all journal data should be committed
+        to the underlying database.
         """
+        
         self._validate_checkpoint(checkpoint)
-        self.journal.commit_checkpoint(checkpoint)
+        journal_data = self.journal.commit_checkpoint(checkpoint)
+
+        if self.journal.is_empty():
+            for key, value in journal_data.items():
+                if value is not None:
+                    self.wrapped_db[key] = value
+                else:
+                    try:
+                        del self.wrapped_db[key]
+                    except KeyError:
+                        pass
+
+
+    def commit_all(self):
+        """
+        Collapses all outstanding changes
+        """
+        while not self.journal.is_empty():
```

Comment:
I think that exposing a `journal.root_checkpoint_id` would be appropriate here so that you could do a single call to `self.commit(self.journal.root_checkpoint_id)`

Reply:
Good point. I wrote that API before I understood commits always merge in subsequent commits. :thumbsup: